bib/internal/progress/progress.go

```diff
@@ -0,0 +1,296 @@
+package progress
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/cheggaaa/pb/v3"
+	"github.com/mattn/go-isatty"
+	"github.com/sirupsen/logrus"
+
+	"github.com/osbuild/images/pkg/osbuild"
+)
+
+var (
+	osStderr         io.Writer = os.Stderr
+	isattyIsTerminal           = isatty.IsTerminal
+)
+
+// ProgressBar is an interfacs for progress reporting when there is
+// an arbitrary amount of sub-progress information (like osbuild)
+type ProgressBar interface {
+	// SetProgress sets the progress details at the given "level".
+	// Levels should start with "0" and increase as the nesting
+	// gets deeper.
+	SetProgress(level int, msg string, done int, total int) error
+	// The high-level message that is displayed in a spinner
```

Comment:
I think I need examples here. Both messages are described as high-level. But also the word "Pulse" is used, and I don't understand what it means.

Reply:
I updated the comment(s) and include the following in
the commit message now:
```
The interface is very tailored to the needs of bib but it should
be similar enough to be reusable by `image-builder-cli`.

It contains a top level spinner that gives the current step:
"Make manifest" or "Building image". The the subprogress from
osbuild: the pipelines and the stages from the pipelines as
progress bars and the last line is the latest status message,
usually something from osbuild like "Starting pipeline qcow2".

The spinenr is a bit of a creative choice, it could also be a 
progress bar because we know the steps (manifest+build). But the 
runtime is totally dominated by the build step so having a first level
progress bar is odd as the UI will spend 90% of the time in it.
Plus the spinner gives it more a more "dynamic" appearnce.
```
I hope this makes it a bit clearer.

Reply:
Fwiw, I can provide a version without the spinner and three levels of subprogress instead, it's pretty trivial to do this, I played with it but I don't like it that much TBH, i prefer the spinner.